src/core/maps/cell_occupancy_estimator.h

```diff
@@ -47,6 +56,73 @@ struct Occupancy {
 struct Beam {
   double x_st, y_st; ///< Coordinates of a start of the beam.
   double x_end, y_end; ///< Coordinates where the beam reaches an obstacle.
+
+  //methods
+  bool touches_rect(const Rectangle& bnds) const {
+    return equal(x_st, x_end, bnds.left) || equal(y_st, y_end, bnds.bot) ||
+           equal(x_st, x_end, bnds.right) || equal(y_st, y_end, bnds.top);
+  }
+
+  Beam operator- () const {
+    return (*this).revert();
+  }
+
+
+
+  bool intersects(const Rectangle& bnds) const {
+    bool is_inside_rect = bnds.is_inside(x_st, y_st) ||
+                          bnds.is_inside(x_end, y_end);
+    if (is_inside_rect) {
+      return true;
+    }
+
+    bool result = false;
+    Ray ray(x_st, y_st, x_end - x_st, y_end - y_st);
+    Intersections intrs = ray.find_intersections(bnds);
+    for (auto intersection : intrs) {
+      result |= bnds.is_on_border(intersection.x, intersection.y);
+    }
+    return result;
+  }
+
+  bool encounters(const Rectangle& bnds) const {
+    if(!bnds.is_on_border(x_end, y_end) || touches_rect(bnds)) {
+      return false;
+    }
+    if (equal(x_end, bnds.left) || equal(x_end, bnds.right)) {
+      return are_strictly_ordered(x_st, x_end, bnds.center_x());
+    }
+    if (equal(y_end, bnds.bot) || equal(y_end, bnds.top)) {
+      return are_strictly_ordered(y_st, y_end, bnds.center_y());
+    }
+    return false;
+  }
+
+  bool reaches_border_passing_through(const Rectangle& bnds) const {
+    return bnds.is_on_border(x_end, y_end) && intersects(bnds);
+  }
+  Beam revert() const {
+    Beam result(*this);
+    std::swap(result.x_st, result.x_end);
+    std::swap(result.y_st, result.y_end);
+    return result;
+  }
+
+  Beam move_start_out(const Rectangle& bnds) const {
```

Comment:
"move_beam_start_out" sounds clearer to me. Taking into account a new beam creation, I recommend the "clone_with_start_outside" name.